File: Examples/intel_examples/examples_core_c/cblas/source/mkl_jit_create_zgemmx.c
/*******************************************************************************
* Copyright 1999-2020 Intel Corporation.
*
* This software and the related documents are Intel copyrighted  materials,  and
* your use of  them is  governed by the  express license  under which  they were
* provided to you (License).  Unless the License provides otherwise, you may not
* use, modify, copy, publish, distribute,  disclose or transmit this software or
* the related documents without Intel's prior written permission.
*
* This software and the related documents  are provided as  is,  with no express
* or implied  warranties,  other  than those  that are  expressly stated  in the
* License.
*******************************************************************************/

/*
!  Content:
!      MKL_JIT_CREATE_ZGEMM
!      Example Program Text ( C Interface )
!******************************************************************************/

#include <stdio.h>
#include <stdlib.h>
#include "mkl.h"

#include "mkl_example.h"
#include "mkl_cblas.h"

int main(int argc, char *argv[])
{
      FILE *in_file;
      char *in_file_name;

      MKL_INT         m, n, k;
      MKL_INT         lda, ldb, ldc;
      MKL_INT         rmaxa, cmaxa, rmaxb, cmaxb, rmaxc, cmaxc;
      MKL_Complex16   alpha, beta;
      MKL_Complex16   *a, *b, *c;
      MKL_LAYOUT      layout;
      MKL_TRANSPOSE   transA, transB;
      MKL_INT         ma, na, mb, nb;
      void*           jitter;
      zgemm_jit_kernel_t  kernel;
      mkl_jit_status_t status;

      printf("\n     M K L _ J I T _ C R E A T E _ Z G E M M  EXAMPLE PROGRAM\n");

/*       Get input parameters                                  */

      if (argc == 1) {
          printf("\n Input data file must be specified as the first parameter");
          return 1;
      }
      in_file_name = argv[1];

/*       Get input data                                        */

      if ((in_file = fopen(in_file_name, "r")) == NULL) {
          printf("\n ERROR opening '%s' with mode=\"r\"\n", in_file_name);
          return 1;
      }
      if (GetIntegerParameters(in_file, &m, &n, &k) != 3) {
          printf("\n ERROR reading m, n, k\n");
          fclose(in_file);
          return 1;
      }
      if (GetScalarsZ(in_file, &alpha, &beta) != 2) {
          printf("\n ERROR reading alpha, beta\n");
          fclose(in_file);
          return 1;
      }
      if( GetCblasCharParameters(in_file, (CBLAS_TRANSPOSE*) &transA, (CBLAS_TRANSPOSE*) &transB, (CBLAS_LAYOUT*) &layout) != 3 ) {
          printf("\n ERROR reading transA, transB, layout\n");
          fclose(in_file);
          return 1;
      }

      if( transA == MKL_NOTRANS ) {
         rmaxa = m + 1;
         cmaxa = k;
         ma    = m;
         na    = k;
      } else {
         rmaxa = k + 1;
         cmaxa = m;
         ma    = k;
         na    = m;
      }
      if( transB == MKL_NOTRANS ) {
         rmaxb = k + 1;
         cmaxb = n;
         mb    = k;
         nb    = n;
      } else {
         rmaxb = n + 1;
         cmaxb = k;
         mb    = n;
         nb    = k;
      }
      rmaxc = m + 1;
      cmaxc = n;
      a = (MKL_Complex16 *) mkl_calloc(rmaxa*cmaxa, sizeof(MKL_Complex16), 64);
      b = (MKL_Complex16 *) mkl_calloc(rmaxb*cmaxb, sizeof(MKL_Complex16), 64);
      c = (MKL_Complex16 *) mkl_calloc(rmaxc*cmaxc, sizeof(MKL_Complex16), 64);
      if(!a || !b || !c) {
          printf( "\n Can't allocate memory for arrays\n");
          mkl_free(a);
          mkl_free(b);
          mkl_free(c);
          fclose(in_file);
          return 1;
      }
      if( layout == MKL_ROW_MAJOR ) {
          lda=cmaxa;
          ldb=cmaxb;
          ldc=cmaxc;
      } else {
          lda=rmaxa;
          ldb=rmaxb;
          ldc=rmaxc;
      }
      if( GetArrayZ(in_file, (CBLAS_LAYOUT*) &layout, GENERAL_MATRIX, &ma, &na, a, &lda) != 0 ) {
        printf("\n ERROR reading array A\n");
        mkl_free(a);
        mkl_free(b);
        mkl_free(c);
        fclose(in_file);
        return 1;
      }
      if( GetArrayZ(in_file, (CBLAS_LAYOUT*) &layout, GENERAL_MATRIX, &mb, &nb, b, &ldb) != 0 ) {
        printf("\n ERROR reading array B\n");
        mkl_free(a);
        mkl_free(b);
        mkl_free(c);
        fclose(in_file);
        return 1;
      }
      if( GetArrayZ(in_file, (CBLAS_LAYOUT*) &layout, GENERAL_MATRIX, &m,  &n,  c, &ldc) != 0 ) {
        printf("\n ERROR reading array C\n");
        mkl_free(a);
        mkl_free(b);
        mkl_free(c);
        fclose(in_file);
        return 1;
      }
      fclose(in_file);

/*       Print input data                                      */

      printf("\n     INPUT DATA");
      printf("\n       M="INT_FORMAT"  N="INT_FORMAT"  K="INT_FORMAT, m, n, k);
      printf("\n       ALPHA=(%5.1f,%5.1f)  BETA=(%5.1f,%5.1f)", alpha.real, alpha.imag, beta.real, beta.imag);
      PrintParameters("TRANSA, TRANSB", transA, transB);
      PrintParameters("LAYOUT", layout);
      PrintArrayZ((CBLAS_LAYOUT*) &layout, FULLPRINT, GENERAL_MATRIX, &ma, &na, a, &lda, "A");
      PrintArrayZ((CBLAS_LAYOUT*) &layout, FULLPRINT, GENERAL_MATRIX, &mb, &nb, b, &ldb, "B");
      PrintArrayZ((CBLAS_LAYOUT*) &layout, FULLPRINT, GENERAL_MATRIX, &m,  &n,  c, &ldc, "C");

/*      Call MKL_JIT_CREATE_ZGEMM subroutine ( C Interface )                  */

      status = mkl_jit_create_zgemm(&jitter, layout, transA, transB, m, n, k, &alpha,
                                    lda, ldb, &beta, ldc);
      
      if (status == MKL_JIT_ERROR) {
          printf("Error: cannot create jitter\n");
          return 1;
      }

      kernel = mkl_jit_get_zgemm_ptr(jitter);

      kernel(jitter, a, b, c);

/*       Print output data                                     */

      printf("\n\n     OUTPUT DATA");
      PrintArrayZ((CBLAS_LAYOUT*) &layout, FULLPRINT, GENERAL_MATRIX, &m, &n, c, &ldc, "C");

      mkl_jit_destroy(jitter);

      mkl_free(a);
      mkl_free(b);
      mkl_free(c);

      return 0;
}

File: Examples/intel_examples/examples_core_f/blas/source/mkl_jit_create_zgemmx.f90
!===============================================================================
! Copyright 1999-2020 Intel Corporation.
!
! This software and the related documents are Intel copyrighted  materials,  and
! your use of  them is  governed by the  express license  under which  they were
! provided to you (License).  Unless the License provides otherwise, you may not
! use, modify, copy, publish, distribute,  disclose or transmit this software or
! the related documents without Intel's prior written permission.
!
! This software and the related documents  are provided as  is,  with no express
! or implied  warranties,  other  than those  that are  expressly stated  in the
! License.
!===============================================================================

!  Content:
!      MKL_JIT_CREATE_ZGEMM  Example Program Text
!*******************************************************************************

include "mkl_blas.f90"

      program   MKL_JIT_CREATE_ZGEMM_MAIN
!

      use, intrinsic :: ISO_C_BINDING, only : c_ptr, c_int, c_intptr_t, c_funptr, c_double, c_f_procpointer
      use MKL_JIT_BLAS

      TYPE(C_PTR)              jitter
      character*1              transa, transb
      integer(KIND=c_intptr_t) status
      integer                  m, n, k
      integer                  lda, ldb, ldc
      complex(KIND=c_double)   alpha, beta
      integer                  rmaxa, cmaxa, rmaxb, cmaxb, rmaxc, cmaxc
      parameter                (rmaxa=4, cmaxa=4, rmaxb=5, cmaxb=5, rmaxc=4, cmaxc=5)
      parameter                (lda=rmaxa, ldb=rmaxb, ldc=rmaxc)
      complex(KIND=c_double)   a(rmaxa,cmaxa), b(rmaxb,cmaxb), c(rmaxc,cmaxc)
      TYPE(C_FUNPTR)           c_kernel

      PROCEDURE(zgemm_jit_kernel_t),POINTER :: f_kernel

!       External Subroutines
      external         PrintArrayZ

!
!      Executable Statements
!
      print*
      print*,'   MKL_JIT_CREATE_ZGEMM EXAMPLE PROGRAM'

!
!      Read input data
      read*
      read*, m, n, k
      read*, alpha, beta
      read 100, transa, transb
      if ((transa.eq.'N').or.(transa.eq.'n')) then
        if (m.gt.rmaxa.or.k.gt.cmaxa) then
          print*, ' Insufficient memory for arrays'
          goto 999
        end if
        read*, ((a(i,j),j=1,k),i=1,m)
      else
        if (k.gt.rmaxa.or.m.gt.cmaxa) then
          print*, ' Insufficient memory for arrays'
          goto 999
        end if
        read*, ((a(i,j),j=1,m),i=1,k)
      end if
      if ((transb.eq.'N').or.(transb.eq.'n')) then
        if (k.gt.rmaxb.or.n.gt.cmaxb) then
          print*, ' Insufficient memory for arrays'
          goto 999
        end if
        read*, ((b(i,j),j=1,n),i=1,k)
      else
        if (n.gt.rmaxb.or.k.gt.cmaxb) then
          print*, ' Insufficient memory for arrays'
          goto 999
        end if
        read*, ((b(i,j),j=1,k),i=1,n)
      end if
      if (m.gt.rmaxc.or.n.gt.cmaxc) then
          print*, ' Insufficient memory for arrays'
          goto 999
      end if
      read*, ((c(i,j),j=1,n),i=1,m)
!
!       Print input data
      print*
      print*, '     INPUT DATA'
      print 101, m, n, k
      print 102, alpha, beta
      print 103, transa, transb
      if ((transa.eq.'N').or.(transa.eq.'n')) then
        call PrintArrayZ(0,0,m,k,a,lda,'A')
      else
        call PrintArrayZ(0,0,k,m,a,lda,'A')
      end if
      if ((transb.eq.'N').or.(transb.eq.'n')) then
        call PrintArrayZ(0,0,k,n,b,ldb,'B')
      else
        call PrintArrayZ(0,0,n,k,b,ldb,'B')
      end if
      call PrintArrayZ(0,0,m,n,c,ldc,'C')
!
!      Call MKL_JIT_CREATE_ZGEMM subroutine

      status = MKL_JIT_CREATE_ZGEMM(jitter,transa,transb,m,n,k,alpha,lda,ldb,beta,ldc)

      if ((status.eq.MKL_JIT_ERROR)) then
         print*, 'jit error'
         goto 999
      end if

      c_kernel = mkl_jit_get_zgemm_ptr(jitter)

      ! convert C function pointer to fortran procedure pointer
      CALL c_f_procpointer (c_kernel, f_kernel)

      call f_kernel(jitter, a, b, c)

      ! free memory
      status = mkl_jit_destroy(jitter)

!
      print*
      print*, '     OUTPUT DATA'
      call PrintArrayZ(1,0,m,n,c,ldc,'C')

      stop
 100  format(2(a1,1x))
 101  format(7x,'M=',i1,'  N=',i1,'  K=',i1)
 102  format(7x,'ALPHA=(',f5.2,',',f5.2,' )', &
               '  BETA=(',f5.2,',',f5.2,' )')
 103  format(7x,'TRANSA=',a1, '  TRANSB=',a1)
 999  stop 1
      end
